pkg/ingest/conn.go

```diff
@@ -172,3 +175,72 @@ func (m *connectionManager) isEmpty() bool {
 	defer m.mtx.RUnlock()
 	return len(m.active) <= 0
 }
+
+// streamClock is a logical clock that is initialized to start at a random offset
+// that identifies a stream.
+// It always holds the next clock tick and is incremented when Read is called.
+// It's guaranteed that the 4 high bytes are 0 initially. If they overflow, the random
+// component is incremented by 1, effectively creating a new stream identifier.
+// It is not safe to be used concurrently.
+type streamClock [10]byte
+
+func newStreamClock() (*streamClock, error) {
+	var sc streamClock
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	if _, err := r.Read(sc[:6]); err != nil {
+		return nil, err
+	}
+	// Set a bit of the random part to 0 so we can increment the random part
+	// when the clock overflows.
+	// We pick the first bit of the second byte as it simplifies the Read implementation.
+	sc[2] &= 0x7f
+	return &sc, nil
+}
+
+// Read populates b with the next clock tick and increments the clock.
+// b must be of length 10.
+func (sc *streamClock) Read(b []byte) (int, error) {
+	if len(b) != 10 {
+		return 0, fmt.Errorf("illegal read of length %d, expected 10", len(b))
+	}
+	lc := binary.BigEndian.Uint64(sc[2:])
+	copy(b, sc[:])
+
+	binary.BigEndian.PutUint64(sc[2:], lc+1)
+
+	return len(b), nil
+}
+
+type streamClock2 struct {
+	prefix [2]byte
+	clock  uint64
+}
+
+func newStreamClock2() (*streamClock2, error) {
+	var sc streamClock2
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	if _, err := r.Read(sc.prefix[:]); err != nil {
+		return nil, err
+	}
+	// Set a bit of the random part to 0 so we can increment the random part
+	// when the clock overflows.
+	sc.clock = uint64(r.Int31()) << 32
+
+	return &sc, nil
+}
+
+// Read populates b with the next clock tick and increments the clock.
+// b must be of length 10.
+func (sc *streamClock2) Read(b []byte) (int, error) {
+	if len(b) != 10 {
+		return 0, fmt.Errorf("illegal read of length %d, expected 10", len(b))
+	}
+	copy(b, sc.prefix[:])
+	binary.BigEndian.PutUint64(b[2:], sc.clock)
+
+	sc.clock++
```

Comment:
With this alternative implementation, if necessary, we can use atomic increments to make this clock safe for concurrent use. @peterbourgon would know better if there's a use case for such.

Reply:
Yea, but since we need to serialise the disk writes for the respective records to not get the wrong order after all, I don't think we'd benefit from that.

Reply:
The stream clock would be used during per-record ULID generation in the ingester, which is naturally synchronized via the bufio.Scanner driving the writes. So I think this is OK.